Feature/api

lucblender opened the pull request:
Added routes:
- getAccessibleAttributes 
- getOwnedEndpoints
- getUserList
- modifyUserPassword
- addUserAuthority
- removeUserAuthority
Plenty of minor fixes!

lucblender committed:
Minor fix on api, and new routes

added routes getAccessibleAttributes, getOwnedEndpoints, /api/v1/getUserList

lucblender committed:
add some user management routes

modifyUserPassword, addUserAuthority, removeUserAuthority

lucblender committed:
Fix of some access control issue in API

 - if user is endpoint owner
- how we get endpoint permission
- if we want to set endpoint -> right above read
- add friendlyName for endpoint retrieval

lucblender committed:
Clean up and small fix

cm0x4D commented on src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
     fun getNode(endpointEntityRepository: EndpointEntityRepository, nodeRequest: NodeRequest): Node? {
-        val splittedTopic = nodeRequest.nodeTopic.split("/")
-        return endpointEntityRepository.findByIdOrNull(splittedTopic[0])?.endpoint?.nodes?.get(splittedTopic[1])
+        val splitTopic = nodeRequest.nodeTopic.split("/")
```
Check if the split operation gives at least 2 parts before accessing [0] and [1]

lucblender replied:
Fixed in 0b262d6, added CloudioApiException and so removed the ApiActionAnswer data class

cm0x4D commented on src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
     }
 
     fun getWotNode(endpointEntityRepository: EndpointEntityRepository, nodeRequest: NodeRequest, host: String): WotNode? {
+        val splitTopic = nodeRequest.nodeTopic.split("/")
```
Check if the split operation gives at least 2 parts before accessing [0] and [1]

lucblender replied:
Fixed in 0b262d6, same comment as above

cm0x4D commented on src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
-        val splittedTopic = Stack<String>()
-        splittedTopic.addAll(objectRequest.objectTopic.split("/").toList().reversed())
-        val endpointEntity = endpointEntityRepository.findByIdOrNull(splittedTopic.pop())
+        val splitTopic = Stack<String>()
```
Handle EmptyStackException

lucblender replied:
Fixed in 0b262d6, same comment as above

cm0x4D commented on src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
-        val splittedTopic = Stack<String>()
-        splittedTopic.addAll(attributeRequest.attributeTopic.split("/").toList().reversed())
-        val endpointEntity = endpointEntityRepository.findByIdOrNull(splittedTopic.pop())
+        val splitTopic = Stack<String>()
```
Handle EmptyStackException

lucblender replied:
Fixed in 0b262d6, same comment as above

cm0x4D commented on src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:
```diff
@@ -75,16 +83,61 @@ object  EndpointManagementUtil{
             return ApiActionAnswer(false, "Attribute is nor a SetPoint, neither a Parameter")
         else {
+            //send message to amq.topic queue
             rabbitTemplate.convertSendAndReceive("amq.topic",
```
If you do not expect an answer, use the convertAndSend() method.

lucblender replied:
Fixed in 0b262d6

lucblender committed:
check split topic size, handle EmptyStackException and convertAndSend

Since check split size and EmptyStackException are linked to exception I did create a CloudioApiException class for it and used this exception to remplace the class ApiActionAnswer that was acting kinda like an exception
